riscv/src/register/macros.rs

```diff
@@ -397,3 +397,681 @@ macro_rules! clear_pmp {
+#[macro_export]
+macro_rules! csr {
+    ($doc:expr, $ty:ident: $csr:literal, $mask:literal) => {
+        #[repr(C)]
+        #[doc = $doc]
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub struct $ty {
+            bits: usize,
+        }
+
+        impl $ty {
+            /// Bitmask for legal bitfields of the CSR type.
+            pub const BITMASK: usize = $mask;
+
+            /// Creates a new CSR type from raw bits value.
+            ///
+            /// Only bits in the [BITMASK](Self::BITMASK) will be set.
+            pub const fn from_bits(bits: usize) -> Self {
+                Self { bits: bits & $mask }
+            }
+
+            /// Gets the raw bits value.
+            pub const fn bits(&self) -> usize {
+                self.bits & $mask
+            }
+
+            /// Gets the bitmask for the CSR type's bitfields.
+            pub const fn bitmask(&self) -> usize {
+                Self::BITMASK
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! csr_field_enum {
+    ($field_ty_doc:expr,
+     $field_ty:ident {
+         range: [$field_start:literal : $field_end:literal],
+         default: $default_variant:ident,
+         $($variant:ident = $value:expr$(,)?)+
+     }$(,)?
+    ) => {
+         #[doc = $field_ty_doc]
+         #[repr(usize)]
+         #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+         pub enum $field_ty {
+             $($variant = $value),+
+         }
+
+         impl $field_ty {
+             /// Creates a new field variant.
+             pub const fn new() -> Self {
+                 Self::$default_variant
+             }
+
+             /// Attempts to convert a [`usize`] into a valid variant.
+             pub const fn from_usize(val: usize) -> Option<Self> {
+                 match val {
+                     $($value => Some(Self::$variant),)+
+                     _ => None,
+                 }
+             }
+
+             /// Converts the variant into a [`usize`].
+             pub const fn into_usize(self) -> usize {
+                 self as usize
+             }
+         }
+
+         impl Default for $field_ty {
+             fn default() -> Self {
+                 Self::new()
+             }
+         }
+
+         impl From<$field_ty> for usize {
+             fn from(val: $field_ty) -> Self {
+                 val.into_usize()
+             }
+         }
+    };
+}
+
+/// Helper macro to create a read-write CSR type.
+///
+/// The type allows to read the CSR value into memory, and update the field values in-memory.
+///
+/// The user can then write the entire bitfield value back into the CSR with a single write.
+#[macro_export]
+macro_rules! read_write_csr {
+    ($doc:expr,
+     $ty:ident: $csr:tt,
+     $mask:tt,
+     $(
+     $field_doc:expr,
+     $field:ident: $get_bit:literal,
+     $set_field_doc:expr,
+     $set_field:ident: $set_bit:literal$(,)?
```

Comment:
Another thing I would like your opinion on @romancardenas, do you think there are use-cases where users would want to define different bits for the `get` and `set` functions?

I couldn't really think of a place where this would be a good idea, so maybe the `bit` argument could be a single parameter?

Same idea for the range and multi-bit variants.